### DISCUSSION TITLE
undefined reference to

### DISCUSSION THREAD

**Discussion by ccoqueiro** [open-telemetry/opentelemetry-cpp#1342] [Q&A]
Hello community,

I'm trying to compile my first code with the opentelemetry SDK, but I'm with the error below, has anyone been through this? Can you help? 

usr/bin/ld: /tmp/ccC5ZrEp.o: in function `(anonymous namespace)::InitTracer()':
helloworld_v1.cpp:(.text+0x236): undefined reference to `opentelemetry::v1::exporter::jaeger::JaegerExporter::JaegerExporter(opentelemetry::v1::exporter::jaeger::JaegerExporterOptions const&)'
/usr/bin/ld: helloworld_v1.cpp:(.text+0x2f1): undefined reference to `opentelemetry::v1::sdk::resource::Resource::Create(opentelemetry::v1::sdk::common::AttributeMap const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: helloworld_v1.cpp:(.text+0x3ab): undefined reference to `opentelemetry::v1::sdk::trace::TracerProvider::TracerProvider(std::unique_ptr<opentelemetry::v1::sdk::trace::SpanProcessor, std::default_delete<opentelemetry::v1::sdk::trace::SpanProcessor> >, opentelemetry::v1::sdk::resource::Resource, std::unique_ptr<opentelemetry::v1::sdk::trace::Sampler, std::default_delete<opentelemetry::v1::sdk::trace::Sampler> >, std::unique_ptr<opentelemetry::v1::sdk::trace::IdGenerator, std::default_delete<opentelemetry::v1::sdk::trace::IdGenerator> >)'
/usr/bin/ld: helloworld_v1.cpp:(.text+0x4f8): undefined reference to `opentelemetry::v1::sdk::resource::Resource::Create(opentelemetry::v1::sdk::common::AttributeMap const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: helloworld_v1.cpp:(.text+0x516): undefined reference to `opentelemetry::v1::sdk::resource::Resource::GetAttributes() const'
/usr/bin/ld: /tmp/ccC5ZrEp.o: in function `opentelemetry::v1::sdk::trace::RandomIdGenerator::RandomIdGenerator()':
helloworld_v1.cpp:(.text._ZN13opentelemetry2v13sdk5trace17RandomIdGeneratorC2Ev[_ZN13opentelemetry2v13sdk5trace17RandomIdGeneratorC5Ev]+0x1f): undefined reference to `vtable for opentelemetry::v1::sdk::trace::RandomIdGenerator'
collect2: error: ld returned 1 exit status

My code:

#include "opentelemetry/exporters/jaeger/jaeger_exporter.h"
#include "opentelemetry/sdk/trace/simple_processor.h"
#include "opentelemetry/sdk/trace/tracer_provider.h"
#include "opentelemetry/trace/provider.h"
#include <iostream>

namespace trace     = opentelemetry::trace;
namespace nostd     = opentelemetry::nostd;
namespace trace_sdk = opentelemetry::sdk::trace;
namespace jaeger    = opentelemetry::exporter::jaeger;

namespace
{
opentelemetry::exporter::jaeger::JaegerExporterOptions opts;
void InitTracer()
{
  opts.transport_format  = opentelemetry::exporter::jaeger::TransportFormat::kThriftHttp;
  opts.endpoint = "localhost";
  opts.server_port =  6831;
  auto exporter  = std::unique_ptr<trace_sdk::SpanExporter>(new jaeger::JaegerExporter(opts));
  auto processor = std::unique_ptr<trace_sdk::SpanProcessor>(
      new trace_sdk::SimpleSpanProcessor(std::move(exporter)));
  auto provider =
      nostd::shared_ptr<trace::TracerProvider>(new trace_sdk::TracerProvider(std::move(processor)));
  // Set the global trace provider
  trace::Provider::SetTracerProvider(provider);
  auto resource_attributes = opentelemetry::sdk::resource::ResourceAttributes
    {
        {"service.name", "hello world"},
        {"deployment.environment", "desenv"}
    };
   auto resource = opentelemetry::sdk::resource::Resource::Create(resource_attributes);
  auto received_attributes = resource.GetAttributes();
}
}  // namespace

int main() {
    std::cout << "Hello World!";
    InitTracer();
    return 0;
}

**Comment by owent** [comment]
What's your build system and what's your link options?
Do you link sdk and jaeger exporter?

**Comment by AkshataKulkarni2504** [comment]
hi, I am facing the same issue. can you let me know how you resolved it.

**Reply by lalitb** [reply]
What is the exact error? The above issue was specific to using Jaeger exporter, which has been removed as part of #2031.

**Reply by AkshataKulkarni2504** [reply]
This is the exact error that I am getting on running roll dice with zipkin exporter.

ld: Undefined symbols:
  opentelemetry::v1::exporter::zipkin::ZipkinExporterFactory::Create(opentelemetry::v1::exporter::zipkin::ZipkinExporterOptions const&), referenced from:
      InitTracer() in main.cpp.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [dice-server] Error 1
make[1]: *** [CMakeFiles/dice-server.dir/all] Error 2

**Reply by lalitb** [reply]
And what are the instructions followed to build otel-cpp, and link it to application?

**Reply by AkshataKulkarni2504** [reply]
I modified main.cpp with init tracer for zipkin exporter to get the traces and used cmake .. and cmake --build . command to build it. I am getting this error when I run cmake --build . command. I followed the steps provided here https://opentelemetry.io/docs/languages/cpp/exporters/

**Reply by lalitb** [reply]
How does the CMakeLists.txt of your sample code looks like? Also did you build and install otel-cpp before trying to build your application. If you are on Linux, and the default install location is /usr/local, check the contents as below:

```console
$ ls /usr/local/lib/libopentelemetry_*
/usr/local/lib/libopentelemetry_common.a                    /usr/local/lib/libopentelemetry_exporter_ostream_span.a  /usr/local/lib/libopentelemetry_trace.a
/usr/local/lib/libopentelemetry_exporter_in_memory.a        /usr/local/lib/libopentelemetry_logs.a                   /usr/local/lib/libopentelemetry_version.a
/usr/local/lib/libopentelemetry_exporter_ostream_logs.a     /usr/local/lib/libopentelemetry_metrics.a
/usr/local/lib/libopentelemetry_exporter_ostream_metrics.a  /usr/local/lib/libopentelemetry_resources.a

$ ls /usr/local/include/opentelemetry/
baggage  common  config.h  context  detail  exporters  ext  logs  metrics  nostd  plugin  sdk  sdk_config.h  std  trace  version.h
```

**Reply by AkshataKulkarni2504** [reply]
Yes I have build and installed otel-cpp. That runs fine. I am building it on macOS.
 This is what my CMakeLists.txt looks like: 

project(RollDiceServer)
cmake_minimum_required(VERSION 3.1)
# Set C++ standard (e.g., C++17)
set(CMAKE_CXX_STANDARD 17)
set(project_name roll-dice-server)

# Define your project's source files
set(SOURCES
    main.cpp  # Add your source files here
)
# Create an executable target
add_executable(dice-server ${SOURCES})

set(OATPP_ROOT ../oatpp)
set(OPENTELEMETRY_ROOT ../opentelemetry-cpp)
find_library(OATPP_LIB NAMES liboatpp.a HINTS "${OATPP_ROOT}/build/src/" NO_DEFAULT_PATH)
if (NOT OATPP_LIB)
  message(SEND_ERROR "Did not find oatpp library ${OATPP_ROOT}/build/src")
endif()
# set the path to the directory containing "oatpp" package configuration files
include_directories(${OATPP_ROOT}/src)

include_directories(${OPENTELEMETRY_ROOT}/api/include)
include_directories(${OPENTELEMETRY_ROOT}/sdk/include)
include_directories(${OPENTELEMETRY_ROOT}/sdk/src)
include_directories(${OPENTELEMETRY_ROOT}/exporters/ostream/include)
include_directories(${OPENTELEMETRY_ROOT}/exporters/zipkin/include)
include_directories(${OPENTELEMETRY_ROOT}/ext/include)
include_directories(${OATPP_ROOT}/src/)

find_library(OPENTELEMETRY_COMMON_LIB NAMES libopentelemetry_common.a HINTS "${OPENTELEMETRY_ROOT}/build/sdk/src/common" NO_DEFAULT_PATH)
find_library(OPENTELEMETRY_TRACE_LIB NAMES libopentelemetry_trace.a HINTS "${OPENTELEMETRY_ROOT}/build/sdk/src/trace" NO_DEFAULT_PATH)
find_library(OPENTELEMETRY_EXPORTER_LIB NAMES libopentelemetry_exporter_ostream_span.a HINTS "${OPENTELEMETRY_ROOT}/build/exporters/ostream" NO_DEFAULT_PATH)
find_library(OPENTELEMETRY_RESOURCE_LIB NAMES libopentelemetry_resources.a HINTS "${OPENTELEMETRY_ROOT}/build/sdk/src/resource" NO_DEFAULT_PATH)

if(OPENTELEMETRY_COMMON_LIB AND OPENTELEMETRY_TRACE_LIB AND OPENTELEMETRY_EXPORTER_LIB AND OPENTELEMETRY_RESOURCE_LIB)
  message(STATUS "Found opentelemetry libraries")
else()
  message(SEND_ERROR "Did not find opentelemetry libraries")
endif()

target_link_libraries(dice-server PRIVATE ${OATPP_LIB} ${OPENTELEMETRY_COMMON_LIB} ${OPENTELEMETRY_TRACE_LIB} ${OPENTELEMETRY_EXPORTER_LIB} ${OPENTELEMETRY_RESOURCE_LIB})

**Reply by lalitb** [reply]
You need to build otel-cpp with `WITH_ZIPKIN` cmake option enabled. This will build the otel zipkin exporter. And then link that library while building your `dice-server`.

**Reply by AkshataKulkarni2504** [reply]
I built Otel-cpp with Zipkin option. could you help me with the name of the library that needs to be linked with dice server for zipkin exporter.

**Reply by lalitb** [reply]
It should be `libopentelemetry_exporter_zipkin_trace`

**Reply by AkshataKulkarni2504** [reply]
Thank you. the issue got resolved.

**Reply by lalitb** [reply]
Thanks. Let us know if the instructions here were not sufficient  - https://opentelemetry.io/docs/languages/cpp/getting-started/. The only difference I think was that these instructions were for using the OStream exporter, while you were trying to use the Zipkin exporter.
